File: mobile/src/main/java/net/mm2d/dmsexplorer/view/PhotoActivity.java
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view;

import android.content.res.Configuration;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.support.v4.view.ViewPager.OnPageChangeListener;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MotionEvent;

import net.mm2d.dmsexplorer.EventLogger;
import net.mm2d.dmsexplorer.R;
import net.mm2d.dmsexplorer.Repository;
import net.mm2d.dmsexplorer.databinding.PhotoActivityBinding;
import net.mm2d.dmsexplorer.domain.model.MediaServerModel;
import net.mm2d.dmsexplorer.util.FullscreenHelper;
import net.mm2d.dmsexplorer.view.base.BaseActivity;
import net.mm2d.dmsexplorer.view.view.ViewPagerAdapter;
import net.mm2d.dmsexplorer.viewmodel.PhotoActivityModel;

/**
 * 静止画表示のActivity。
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介 (OHMAE Ryosuke)</a>
 */
public class PhotoActivity extends BaseActivity {
    private FullscreenHelper mFullscreenHelper;
    private PhotoActivityBinding mBinding;
    private PhotoActivityModel mModel;
    private Repository mRepository;
    private MediaServerModel mServerModel;
    private final OnPageChangeListener mOnPageChangeListener = new OnPageChangeListener() {
        @Override
        public void onPageScrolled(
                final int position,
                final float positionOffset,
                final int positionOffsetPixels) {
        }

        @Override
        public void onPageSelected(final int position) {
        }

        @Override
        public void onPageScrollStateChanged(final int state) {
            if (state == ViewPager.SCROLL_STATE_IDLE) {
                onScrollIdle();
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.photo_activity);
        mFullscreenHelper = new FullscreenHelper.Builder(mBinding.getRoot())
                .setTopView(mBinding.toolbar)
                .build();
        mRepository = Repository.get();
        mServerModel = mRepository.getMediaServerModel();
        try {
            mModel = new PhotoActivityModel(this, mRepository);
        } catch (final IllegalStateException ignored) {
            finish();
            return;
        }

        mBinding.setModel(mModel);
        mModel.adjustPanel(this);
        mFullscreenHelper.showNavigation();

        final ViewPagerAdapter pagerAdapter = new ViewPagerAdapter();
        final LayoutInflater inflater = LayoutInflater.from(this);
        pagerAdapter.add(inflater.inflate(R.layout.progress_view, mBinding.viewPager, false));
        pagerAdapter.add(inflater.inflate(R.layout.transparent_view, mBinding.viewPager, false));
        pagerAdapter.add(inflater.inflate(R.layout.progress_view, mBinding.viewPager, false));
        mBinding.viewPager.setAdapter(pagerAdapter);
        mBinding.viewPager.setCurrentItem(1, false);
        mBinding.viewPager.addOnPageChangeListener(mOnPageChangeListener);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mFullscreenHelper.terminate();
    }

    @Override
    public boolean dispatchKeyEvent(final KeyEvent event) {
        if (event.getAction() == KeyEvent.ACTION_UP
                && event.getKeyCode() != KeyEvent.KEYCODE_BACK) {
            mFullscreenHelper.showNavigation();
        }
        return super.dispatchKeyEvent(event);
    }

    @Override
    public boolean dispatchTouchEvent(final MotionEvent ev) {
        final boolean result = super.dispatchTouchEvent(ev);
        mFullscreenHelper.showNavigation();
        return result;
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        if (mModel != null) {
            mModel.adjustPanel(this);
        }
    }

    private void onScrollIdle() {
        final int index = mBinding.viewPager.getCurrentItem();
        if (index == 1) {
            return;
        }
        if (!move(index)) {
            finish();
            return;
        }
        mModel = new PhotoActivityModel(this, mRepository);
        mBinding.setModel(mModel);
        mBinding.viewPager.setCurrentItem(1, false);
        EventLogger.sendPlayContent(true);
    }

    private boolean move(int index) {
        return index == 0 ? mServerModel.selectPreviousEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL)
                : mServerModel.selectNextEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL);
    }
}

File: mobile/src/main/java/net/mm2d/dmsexplorer/view/PhotoActivity.kt
/*
 * Copyright (c) 2017 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.dmsexplorer.view

import android.content.res.Configuration
import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.MotionEvent
import android.widget.FrameLayout
import androidx.core.view.isVisible
import androidx.core.view.updateLayoutParams
import androidx.lifecycle.flowWithLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import net.mm2d.dmsexplorer.R
import net.mm2d.dmsexplorer.Repository
import net.mm2d.dmsexplorer.databinding.PhotoActivityBinding
import net.mm2d.dmsexplorer.domain.model.MediaServerModel
import net.mm2d.dmsexplorer.log.EventLogger
import net.mm2d.dmsexplorer.settings.Settings
import net.mm2d.dmsexplorer.util.FullscreenHelper
import net.mm2d.dmsexplorer.view.base.BaseActivity
import net.mm2d.dmsexplorer.view.view.ViewPagerAdapter
import net.mm2d.dmsexplorer.viewmodel.PhotoActivityModel
import net.mm2d.dmsexplorer.viewmodel.adapter.ImageViewBindingAdapter

/**
 * 静止画表示のActivity。
 *
 * @author [大前良介 (OHMAE Ryosuke)](mailto:ryo@mm2d.net)
 */
class PhotoActivity : BaseActivity() {
    private lateinit var settings: Settings
    private lateinit var fullscreenHelper: FullscreenHelper
    private lateinit var binding: PhotoActivityBinding
    private lateinit var model: PhotoActivityModel
    private lateinit var repository: Repository
    private lateinit var serverModel: MediaServerModel
    private val onPageChangeListener = object : OnPageChangeListener {
        override fun onPageScrolled(
            position: Int,
            positionOffset: Float,
            positionOffsetPixels: Int,
        ) = Unit

        override fun onPageSelected(position: Int) = Unit

        override fun onPageScrollStateChanged(state: Int) {
            if (state == ViewPager.SCROLL_STATE_IDLE) {
                onScrollIdle()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        settings = Settings.get()
        setTheme(settings.themeParams.fullscreenThemeId)
        super.onCreate(savedInstanceState)
        binding = PhotoActivityBinding.inflate(layoutInflater)
        setContentView(binding.root)
        fullscreenHelper = FullscreenHelper(
            window = window,
            rootView = binding.root,
            topView = binding.toolbar,
        )
        repository = Repository.get()
        val serverModel = repository.mediaServerModel
        if (serverModel == null) {
            finish()
            return
        }
        this.serverModel = serverModel
        try {
            model = PhotoActivityModel(this, repository)
        } catch (ignored: IllegalStateException) {
            finish()
            return
        }
        model.update()
        model.adjustPanel(this)
        if (settings.shouldShowPhotoUiOnStart()) {
            fullscreenHelper.showNavigation()
        } else {
            fullscreenHelper.hideNavigationImmediately()
        }

        val inflater = LayoutInflater.from(this)
        binding.viewPager.adapter = ViewPagerAdapter().also {
            it.add(inflater.inflate(R.layout.progress_view, binding.viewPager, false))
            it.add(inflater.inflate(R.layout.transparent_view, binding.viewPager, false))
            it.add(inflater.inflate(R.layout.progress_view, binding.viewPager, false))
        }
        binding.viewPager.setCurrentItem(1, false)
        binding.viewPager.addOnPageChangeListener(onPageChangeListener)
        model.getImageBinaryFlow()
            .flowWithLifecycle(lifecycle)
            .distinctUntilChanged()
            .onEach {
                ImageViewBindingAdapter.setImageBinary(binding.imageView, it)
            }
            .launchIn(lifecycleScope)
        model.getIsLoadingFlow()
            .flowWithLifecycle(lifecycle)
            .distinctUntilChanged()
            .onEach {
                binding.progressBar.isVisible = it
            }
            .launchIn(lifecycleScope)
        binding.toolbar.setBackgroundColor(model.background)
        model.getRightNavigationSizeFlow()
            .flowWithLifecycle(lifecycle)
            .distinctUntilChanged()
            .onEach {
                binding.toolbar.updateLayoutParams<FrameLayout.LayoutParams> {
                    rightMargin = it
                }
            }
            .launchIn(lifecycleScope)
        binding.toolbarBack.setOnClickListener { model.onClickBack() }
        model.getTitleFlow()
            .flowWithLifecycle(lifecycle)
            .distinctUntilChanged()
            .onEach {
                binding.toolbarTitle.text = it
            }
            .launchIn(lifecycleScope)
    }

    override fun onDestroy() {
        super.onDestroy()
        fullscreenHelper.terminate()
    }

    override fun updateOrientationSettings() {
        settings.photoOrientation
            .setRequestedOrientation(this)
    }

    override fun dispatchKeyEvent(event: KeyEvent): Boolean {
        if (event.action == KeyEvent.ACTION_UP && event.keyCode != KeyEvent.KEYCODE_BACK) {
            fullscreenHelper.showNavigation()
        }
        return super.dispatchKeyEvent(event)
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        val result = super.dispatchTouchEvent(ev)
        if (settings.shouldShowPhotoUiOnTouch()) {
            fullscreenHelper.showNavigation()
        }
        return result
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        model.adjustPanel(this)
    }

    private fun onScrollIdle() {
        val index = binding.viewPager.currentItem
        if (index == 1) {
            return
        }
        if (!move(index)) {
            finish()
            return
        }
        model.update()
        binding.viewPager.setCurrentItem(1, false)
        EventLogger.sendPlayContent(true)
    }

    private fun move(index: Int): Boolean {
        return if (index == 0) {
            serverModel.selectPreviousEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL)
        } else {
            serverModel.selectNextEntity(MediaServerModel.SCAN_MODE_SEQUENTIAL)
        }
    }
}
